OCPSTRAT-898 - Need to ensure ICSP and IDMS interoperate especially in multi-tenant scenario

mburke5678 opened the pull request:
Because the ICSP functionality cannot be removed from the API in 4.x, IDMS/ITMS objects can now exist with ICSP. This PR removes the requirement to delete ICMPs before using IDMS/ITMS objects, replaces references to ICSP in places where [they were removed](https://github.com/openshift/openshift-docs/pull/56250/files). 

[OSDOCS-8697 - update Openshift document about migration details](https://issues.redhat.com/browse/OSDOCS-8697)
[OCPSTRAT-898 - Need to ensure ICSP and IDMS interoperate especially in multi-tenant scenario](https://issues.redhat.com/browse/OCPSTRAT-898)


Previews:
CI/CD -> Builds using BuildConfig -> Creating build inputs -> Builds using BuildConfig -> [Image source](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/cicd/builds/creating-build-inputs#builds-image-source_creating-build-inputs) -- Replaced text with _you can use only global pull secrets_ snippet.

Images -> Image configuration resources -> [Configuring image registry repository mirroring](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/openshift_images/image-configuration#images-configuration-registry-mirror_image-configuration) -- Other than the following, the text is unchanged:
* Removed conceptual material and moved to a new module
* Removed prereq about `ImageContentSourcePolicy` objects 
* Added ICSP references/example to step 3 
* Replaced text with _you can use only global pull secrets_ snippet.

[Converting ImageContentSourcePolicy (ICSP) files for image registry repository mirroring
](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/post_installation_configuration/preparing-for-users#images-configuration-registry-mirror-convert_post-install-preparing-for-users) -- Added two sentences:
* Because the migration results 
* Step 3

[Understanding image registry repository mirroring](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/post_installation_configuration/preparing-for-users#images-configuration-registry-mirror_post-install-preparing-for-users) -- New concept module based largely on existing text. 
* Some updated to text to put info on ICSP into the docs.
* Moved prereq and procedure to new module.
* Added final two paragraphs.
* Added the  _you can use only global pull secrets_ snippet.

Architecture -> Control plane architecture -> [About the Machine Config Operator](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/architecture/control-plane#about-machine-config-operator_control-plane) - Updated the _The following modifications do not trigger a node reboot_ snippet. 

Renamed _Configuring image registry repository mirroring_ to _Understanding image registry repository mirroring_ and promoted to H2 level. Added new _Configuring image registry repository mirroring_ procedural module. For the following assemblies:
[Preparing for users](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/post_installation_configuration/preparing-for-users#images-configuration-registry-mirror_post-install-preparing-for-users)  [Current docs](https://docs.openshift.com/container-platform/4.14/post_installation_configuration/preparing-for-users.html#images-configuration-registry-mirror_post-install-preparing-for-users).
[Updating a cluster in a disconnected environment without the OpenShift Update Service
](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/updating/updating_a_cluster/updating_disconnected_cluster/disconnected-update#images-configuration-registry-mirror_updating-restricted-network-cluster) [Current docs](https://docs.openshift.com/container-platform/4.14/updating/updating_a_cluster/updating_disconnected_cluster/disconnected-update.html#images-configuration-registry-mirror_updating-restricted-network-cluster).
[Image configuration resources](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/openshift_images/image-configuration#images-configuration-registry-mirror_image-configuration)  [Current docs](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-registry-mirror_image-configuration).

QE review:
- [ ] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

QiWang19 reviewed (commented):
Can we add to openshift docs somewhere about migration direction and steps for current ICSP users?  [Recommendation on the migration path to IDMS](https://docs.google.com/document/d/1wCw5pIVyLTzbiUYn0xmd30Y9NrxIyzbqWpC-Cu-1Ncs/edit?usp=sharing) in the Migration Scenarios section.

mburke5678 commented:
@QiWang19 

> Can we add to openshift docs somewhere about migration direction and steps for current ICSP users? Recommendation on the migration path to IDMS in the Migration Scenarios section.

Absolutely. Thanks for the suggestion.

mburke5678 commented:
@QiWang19 We have a NOTE in the [Configuring image registry repository mirroring](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/openshift_images/image-configuration#images-configuration-registry-mirror_image-configuration) section that states ICSP is deprecated, but still supported, and introduces the `oc adm migrate icsp` command to convert ICSP to IDMS. Th note includes a reference to the [Converting ImageContentSourcePolicy (ICSP) files for image registry repository mirroring](https://70881--ocpdocs-pr.netlify.app/openshift-enterprise/latest/openshift_images/image-configuration#images-configuration-registry-mirror-convert_image-configuration) section, which describes the migration from ICSP to IDMS. 
Does this meet your expectations?

Also, we have this line in the docs:
[You cannot add the same repository to both an `ImageDigestMirrorSet` and an `ImageTagMirrorSet` object.](https://github.com/openshift/openshift-docs/pull/70881/files#diff-5fa1176e0fd07350c91b903bf4d7eb0f9fa8eaa37343148a503fcb82823b6da6R318)

Is this true? The migration doc says IDMS and ITMS can use the same **registry**.

QiWang19 commented on modules/images-configuration-registry-mirror-configuring.adoc:
```diff
@@ -0,0 +1,274 @@
+* The main registry is only used if no other mirror works.
+* From the system context, the `Insecure` flags are used as fallback.
+* The format of the `/etc/containers/registries.conf` file has changed recently. It is now version 2 and in TOML format.
+* You cannot add the same repository to both an `ImageDigestMirrorSet` and an `ImageTagMirrorSet` object.
```
> Is this true? The migration doc says IDMS and ITMS can use the same registry.

It's true. The same mirror can stay in `ImageDigestMirrorSet` and an `ImageTagMirrorSet`. Whether the mirror will be used during the pull depends on how the pull spec is specified. We can remove this line from the doc.

QiWang19 commented on modules/images-configuration-registry-mirror.adoc:
```diff
@@ -29,276 +29,30 @@ By pulling container images needed by {product-title} and then bringing those im
-* From the system context, the `Insecure` flags are used as fallback.
-* The format of the `/etc/containers/registries.conf` file has changed recently. It is now version 2 and in TOML format.
-* You cannot add the same repository to both an `ImageDigestMirrorSet` and an `ImageTagMirrorSet` object.
+If you upgraded a cluster, any existing ICSP objects remain stable, and both IDMS and ICSP objects are supported. Workloads using ICSP objects continue to function as expected. However, if you want to take advantage of the fallback policies introduced in the IDMS CRs, you can migrate current workloads to IDMS objects by using the `oc adm migrate icsp` command as shown in the *Converting ImageCont ...
```
Under the [Converting ICSP](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-registry-mirror-convert_image-configuration), we can add the step 3 mentioning that the customer can remove existing ICSP objects after IDMS objects get rolled out. And This will result in the same registries.conf and hence not result in a reboot.

mburke5678 commented:
@arnaik-rh PTAL

arnaik-rh commented:
@lyman9966 PTAL

mburke5678 commented on modules/images-configuration-registry-mirror-convert.adoc:
```diff
@@ -12,6 +12,8 @@ Using an `ImageContentSourcePolicy` (ICSP) object to configure repository mirror
 
 ICSP objects are being replaced by `ImageDigestMirrorSet` and `ImageTagMirrorSet` objects to configure repository mirroring. If you have existing YAML files that you used to create `ImageContentSourcePolicy` objects, you can use the `oc adm migrate icsp` command to convert those files to an `ImageDigestMirrorSet` YAML file. The command updates the API to the current version, changes the `kind` va ...
 
+Because the migration results in the same `registries.conf` file, the cluster does not need to reboot. 
```
@QiWang19 Would changing this line to the following be accurate?

> Because the migration does not change the `registries.conf` file, the cluster does not need to reboot.

QiWang19 replied:
yes, that's correct.

mburke5678 commented:
@QiWang19 @lyman9966 Thank you for your help getting the PR completed. Your quick and detailed comments are very helpful.

lyman9966 commented:
/lgtm

mletalie reviewed (commented):
/label peer-review-in-progress

mletalie commented on modules/images-configuration-registry-mirror.adoc:
```diff
@@ -29,276 +29,30 @@ By pulling container images needed by {product-title} and then bringing those im
-** `ImageTagMirrorSet`. This CR allows you to pull images from a mirrored registry by using image tags.
+** `ImageTagMirrorSet` (ITMS). This object allows you to pull images from a mirrored registry by using image tags. The ITMS CR enables you to set a fall back policy that allows or stops continued attempts to pull from the source registry if the image pull fails.
++
+** `ImageContentSourcePolicy` (ICSP). This object allows you to pull images from a mirrored registry by using digest specifications. An ICSP always falls back to the source registry if the mirrors do not work.
```
```suggestion
** `ImageContentSourcePolicy` (ICSP). This object allows you to pull images from a mirrored registry by using digest specifications. The ICSP CR always falls back to the source registry if the mirrors do not work.
```

mletalie reviewed (commented):
Hello @mburke5678,
   Great job on this doc! Thanks for providing such specific instructions in the PR description, immensely helpful. Just had a couple of questions/suggestions. 
I wasn't sure about "Added ICSP references/example to step 3" but I may have missed it...can you please take a look to make sure it is how you wanted it? Also, just to be clear, you added the "current docs" link to show how the topic is currently presented? Sorry about the confusion on my end. 
Again, great job! 
![image](https://github.com/openshift/openshift-docs/assets/122639474/c5111dab-6119-4fbc-b27f-acd152b4b69d)

mletalie reviewed (commented):
/label peer-review-done
/remove-label peer-review-in-progress
/remove-label peer-review-needed
/lgtm

mburke5678 commented:
"Added ICSP references/example to step 3" -- I added back a code block describing the ICSP object (we removed it some time back when the IDMS was added): 
https://github.com/openshift/openshift-docs/pull/70881/files#diff-27a101211672fed423ee44a9a33b205e3f03d0c45f9e6522c230de0a83cb70edR99-R117

And updated this sentence to include ICSP:
https://github.com/openshift/openshift-docs/pull/70881/files#diff-27a101211672fed423ee44a9a33b205e3f03d0c45f9e6522c230de0a83cb70edR126

"you added the 'current docs' link to show how the topic is currently presented?" Correct.

mletalie commented:
> @mletalie In answer to your questions:
> 
> "Added ICSP references/example to step 3" -- I added back a code block describing the ICSP object (we removed it some time back when the IDMS was added): https://github.com/openshift/openshift-docs/pull/70881/files#diff-27a101211672fed423ee44a9a33b205e3f03d0c45f9e6522c230de0a83cb70edR99-R117
> 
> And updated this sentence to include ICSP: https://github.com/openshift/openshift-docs/pull/70881/files#diff-27a101211672fed423ee44a9a33b205e3f03d0c45f9e6522c230de0a83cb70edR126
> 
> "you added the 'current docs' link to show how the topic is currently presented?" Correct.

Thanks @mburke5678!

mburke5678 commented:
/cherrypick enterprise-4.15

mburke5678 commented:
/cherrypick enterprise-4.14

mburke5678 commented:
/cherrypick enterprise-4.13
